src/v/kafka/server/quota_manager.cc

```diff
@@ -218,100 +210,107 @@ ss::future<clock::duration> quota_manager::maybe_add_and_retrieve_quota(
+
+    co_return it->second;
 }
 
 void quota_manager::update_client_quotas() {
     vassert(
-      ss::this_shard_id() == _client_quotas.shard_id(),
+      ss::this_shard_id() == quotas_shard,
       "update_client_quotas must only be called on the owner shard");
 
+    // Hold to lock to ensure there are no updates to the map while iterating
     ssx::spawn_with_gate(_gate, [this] {
-        return _client_quotas.update([this](client_quotas_map_t quotas) {
-            constexpr auto set_bucket =
-              [](
-                std::optional<atomic_token_bucket>& bucket,
-                std::optional<uint64_t> rate,
-                std::optional<uint64_t> replenish_threshold) {
-                  if (!rate) {
-                      bucket.reset();
-                      return;
-                  }
-                  if (bucket.has_value() && bucket->rate() == rate) {
-                      return;
-                  }
-                  bucket.emplace(
-                    *rate, *rate, replenish_threshold.value_or(1), true);
-              };
-
-            for (auto& quota : quotas) {
-                auto limits = _translator.find_quota_value(quota.first);
-                set_bucket(
-                  quota.second->tp_produce_rate,
-                  limits.produce_limit,
-                  _replenish_threshold());
-
-                set_bucket(
-                  quota.second->tp_fetch_rate,
-                  limits.fetch_limit,
-                  _replenish_threshold());
-
-                set_bucket(
-                  quota.second->pm_rate,
-                  limits.partition_mutation_limit,
-                  _replenish_threshold());
-            }
-
-            return quotas;
-        });
+        return _global_map_mutex->with(
+          [this] { return do_update_client_quotas(); });
     });
 }
 
+ss::future<> quota_manager::do_update_client_quotas() {
+    constexpr auto set_bucket = [](
+                                  std::optional<atomic_token_bucket>& bucket,
+                                  std::optional<uint64_t> rate,
+                                  std::optional<uint64_t> replenish_threshold) {
+        if (!rate) {
+            bucket.reset();
+            return;
+        }
+
+        if (bucket.has_value() && bucket->rate() == rate) {
+            return;
+        }
+        bucket.emplace(*rate, *rate, replenish_threshold.value_or(1), true);
+        return;
+    };
+
+    for (auto& quota : *_global_map) {
+        auto limits = _translator.find_quota_value(quota.first);
+        set_bucket(
+          quota.second->tp_produce_rate,
+          limits.produce_limit,
+          _replenish_threshold());
+
+        set_bucket(
+          quota.second->tp_fetch_rate,
+          limits.fetch_limit,
+          _replenish_threshold());
+
+        set_bucket(
+          quota.second->pm_rate,
+          limits.partition_mutation_limit,
+          _replenish_threshold());
+
+        co_await ss::maybe_yield();
```

Comment:
You could use ssx::async_for_each for this loop.

Reply:
Done